executor/insert_common.go

```diff
@@ -792,6 +807,62 @@ func getAutoRecordID(d types.Datum, target *types.FieldType, isInsert bool) (int
 	return recordID, nil
 }
 
+func (e *InsertValues) adjustAutoShardDatum(ctx context.Context, d types.Datum, hasValue bool, c *table.Column) (types.Datum, error) {
+	if !hasValue || d.IsNull() {
```

Comment:
should take retry into consideration.

Reply:
Can we simply ignore the retrying, just allocate another id instead?

Reply:
It's better to cache the auto shard id in `retryInfo`, but we can do that in another PR, after all this PR is very large already.